File: Frontend/src/interfaces/pose.ts
import * as THREE from "three";

export class Pose {
    j_1: THREE.Vector3;
    j_2: THREE.Vector3;
    j_3: THREE.Vector3;
    j_4: THREE.Vector3;
    j_5: THREE.Vector3;
    j_6: THREE.Vector3;
    j_7: THREE.Vector3;
    theta_0: number;
    theta_1: number;
    theta_2: number;
    theta_3: number;

    constructor(jsonString : string) {
        const json = JSON.parse(jsonString);

        this.j_1 = new THREE.Vector3(json['j_1'][0], json['j_1'][2], -json['j_1'][1]);
        this.j_2 = new THREE.Vector3(json['j_2'][0], json['j_2'][2], -json['j_2'][1]);
        this.j_3 = new THREE.Vector3(json['j_3'][0], json['j_3'][2], -json['j_3'][1]);
        this.j_4 = new THREE.Vector3(json['j_4'][0], json['j_4'][2], -json['j_4'][1]);
        this.j_5 = new THREE.Vector3(json['j_5'][0], json['j_5'][2], -json['j_5'][1]);
        this.j_6 = new THREE.Vector3(json['j_6'][0], json['j_6'][2], -json['j_6'][1]);
        this.j_7 = new THREE.Vector3(json['j_7'][0], json['j_7'][2], -json['j_7'][1]);
        this.theta_0 = json['theta_0'];
        this.theta_1 = json['theta_1'];
        this.theta_2 = json['theta_2'];
        this.theta_3 = json['theta_3'];
    }
}

File: backend/app/models/Pose.py
import json

import numpy as np

from backend.app.models.ActuatorStates import ActuatorStates


def convert_to_xyz_tuple(frame: np.ndarray) -> tuple:
    frame_t = frame[0:3, 3:4].T
    return tuple(map(tuple, frame_t))[0]


class Pose(object):
    """
    Defining the robot pose including the origin for rendering it on the frontend
    The pose is represented by the xyz coordinates of all the joints and the rotation angles
    """

    def __init__(self, robot_frames: np.ndarray, origin_t_1: tuple, act_states_t_1: ActuatorStates):
        self.j_1 = convert_to_xyz_tuple(robot_frames[0])
        self.j_2 = convert_to_xyz_tuple(robot_frames[1])
        self.j_3 = convert_to_xyz_tuple(robot_frames[2])
        self.j_4 = convert_to_xyz_tuple(robot_frames[3])
        self.j_5 = convert_to_xyz_tuple(robot_frames[4])
        self.j_6 = convert_to_xyz_tuple(robot_frames[5])
        self.j_7 = convert_to_xyz_tuple(robot_frames[6])

        self.theta_0 = origin_t_1[3]
        self.theta_1 = act_states_t_1.theta_1
        self.theta_2 = act_states_t_1.theta_2
        self.theta_3 = act_states_t_1.theta_3

    def to_json(self) -> json:
        return json.dumps(self.__dict__)
